test/azul_test_case.py

```diff
@@ -3,12 +3,21 @@
 from unittest.mock import (
     patch,
 )
+import warnings
 
 import boto3.session
 from botocore.credentials import Credentials
 import botocore.session
 
 
+class AzulTestCase(TestCase):
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        super().setUpClass()
+        warnings.filterwarnings('ignore', 'unclosed.*<ssl.SSLSocket.*>', ResourceWarning)
```

Comment:
Every action in `setUp` should be matched by an inverse action in `tearDown`. IOW, `tearDown` should leave the world in the state that `setUp` was called in.

You could use `warning.resetwarnings()` but then you would have to assert in `setUp` that the warnings are indeed at the reset state. Luckily, there is `warnings.catch_warnings` which was made exactly for this purpose. You can even assert in `tearDown` that only expected warnings occurred. You can't use `warnings.catch_warnings` in conjunction with a `with` statement because a statement can't be split over two methods, but you can call `__enter__` and `__exit__` explicitly. Be sure to read the [PEP on that](https://www.python.org/dev/peps/pep-0343/#specification-the-with-statement). Plus there are some fixtures in our tests that use that technique already.

Reply:
Now that we use the `catch_warnings` context manager we can actually assert that we only get a specific set of warnings. It's possible that we get no warnings or a subset of the set of expected warnings but the test should fail if we get an unexpected warning.

https://docs.python.org/3/library/warnings.html#testing-warnings

That's what I meant by 

> You can even assert in tearDown that only expected warnings occurred.

in my previous comment on this.

Reply:
@hannes-ucsc is there a list of Exceptions that we would permit to pass through?

Reply:
During an IT with a filter of: 

`warnings.filterwarnings('ignore', 'unclosed.*<ssl.SSLSocket.*>', ResourceWarning)`

the catch_warning record shows multiple instances of :

`{message : DeprecationWarning('Call to deprecated method fetch_bundle_manifest.'), category : 'DeprecationWarning', filename : '/Users/amar/dev/hca/azul/src/azul/azulclient.py', lineno : 92, line : None}`

Reply:
Decided to remove filterwarnings, and inspect all warnings caught by the `catch_warnings` these WarningMessages are to be compared against the permitted warnings, and regular expressions that are defined to finely filter messages we expect to see. 
We have to inspect ever single warning, and need to ensure that ALL warnings that fail to meet the filter requirements are displayed.